pytential/qbx/distributed.py

```diff
@@ -0,0 +1,1071 @@
+            cost_model = QBXCostModel()
+
+            import warnings
+            warnings.warn(
+                "Kernel-specific calibration parameters are not supplied when"
+                "using distributed FMM.")
+            # TODO: supply better default calibration parameters
+            calibration_params = AbstractQBXCostModel.get_unit_calibration_params()
+
+            kernel_args = {}
+            for arg_name, arg_expr in six.iteritems(insn.kernel_arguments):
+                kernel_args[arg_name] = evaluate(arg_expr)
+
+            boxes_time, _ = cost_model.qbx_cost_per_box(
+                actx.queue, global_geo_data_device, insn.target_kernels[0],
+                kernel_args, calibration_params)
+            boxes_time = boxes_time.get()
+
+            from pytential.qbx.utils import ToHostTransferredGeoDataWrapper
+            global_geo_data = ToHostTransferredGeoDataWrapper(global_geo_data_device)
+
+        # FIXME Exert more positive control over geo_data attribute lifetimes using
+        # geo_data.<method>.clear_cache(geo_data).
+
+        # FIXME Synthesize "bad centers" around corners and edges that have
+        # inadequate QBX coverage.
+
+        # FIXME don't compute *all* output kernels on all targets--respect that
+        # some target discretizations may only be asking for derivatives (e.g.)
+
+        # {{{ Construct a traversal builder
+
+        # NOTE: The distributed implementation relies on building the same traversal
+        # objects as the one on the root rank. This means here the traversal builder
+        # should use the same parameters as `QBXFMMGeometryData.traversal`. To make
+        # it consistent across ranks, we broadcast the parameters here.
+
+        trav_param = None
+        if self.comm.Get_rank() == 0:
+            trav_param = {
+                "well_sep_is_n_away":
+                    global_geo_data.geo_data.code_getter.build_traversal
+                    .well_sep_is_n_away,
+                "from_sep_smaller_crit":
+                    global_geo_data.geo_data.code_getter.build_traversal.
+                    from_sep_smaller_crit,
+                "_from_sep_smaller_min_nsources_cumul":
+                    global_geo_data.geo_data.lpot_source.
+                    _from_sep_smaller_min_nsources_cumul}
+        trav_param = self.comm.bcast(trav_param, root=0)
+
+        traversal_builder = QBXFMMGeometryDataTraversalBuilder(
+            actx.context,
+            well_sep_is_n_away=trav_param["well_sep_is_n_away"],
+            from_sep_smaller_crit=trav_param["from_sep_smaller_crit"],
+            _from_sep_smaller_min_nsources_cumul=trav_param[
+                "_from_sep_smaller_min_nsources_cumul"])
+
+        # }}}
+
+        # {{{ Broadcast the subset of the global geometry data to worker ranks
+
+        global_geo_data = broadcast_global_geometry_data(
+            self.comm, actx, traversal_builder, global_geo_data)
+
+        # }}}
```

Comment:
Builds traversal every time the operator is called.

Reply:
Use `GeometryCollection` to cache the geometry data and traversal objects on the worker ranks.

Reply:
The geometry data is cached in the `DistributedBoundExpression` object, [here](https://github.com/gaohao95/pytential/blob/distributed/pytential/symbolic/execution.py#L970).